### DISCUSSION TITLE
How do you make __typename required in type when fetching it explicitly?

### DISCUSSION THREAD

**Discussion by ericbf** [dotansimha/graphql-code-generator#8786] [Q&A]
So I have a union type that I discriminate by the `__typename` property. The problem I’m seeing is that in the generated types for my query where I fetch `__typename` from the union type, `__typename` is optional. It shouldn’t be optional in the return type since I’m expressly getting it through my query:

```ts
query SomeQuery {
    someUnion {
        __typename
    }
}
```

Results in a return type with `__typename` as optional, even though I’m specifically requesting it.

So how do I make the `__typename`s not optional (i.e., required)?

**Comment by saihaj** [comment]
add this to your config [`nonOptionalTypename: true`](https://the-guild.dev/graphql/codegen/plugins/typescript/typescript-operations#usage-examples-24) I think there might be a bug for us to fix here 🤔

**Reply by ericbf** [reply]
That does make them non-optional, but even if they aren’t specified in the query. That is better, but even better would be its being non-optional if I explicitly put it in my query, but optional or even omitted otherwise. Is that possible?

**Reply by saihaj** [reply]
I think it is likely a regression.

**Reply by saihaj** [reply]
that can act as workaround for now until we land a fix here https://github.com/dotansimha/graphql-code-generator/issues/8793

**Reply by saihaj** [reply]
@ericbf what does your config look like I am trying to repro this?

**Reply by ericbf** [reply]
I’ll create a repro

**Reply by ericbf** [reply]
See my comment on the issue: https://github.com/dotansimha/graphql-code-generator/issues/8793#issuecomment-1384331934
